interpreter/src/lib.rs

```diff
@@ -37,44 +39,117 @@ impl State {
         self.memory.insert(x.to_string(), value);
     }
 
-    pub fn write(&mut self, x: &str, value: u32) {
+    pub fn write(&mut self, x: &str, value: u32, thread: &str) {
+
         match self.memory_model {
             MemoryModel::Sc => {
                 self.memory.insert(x.to_string(), value);
             }
             MemoryModel::Tso => {
-                todo!()
+                if let Some(buffer) = self.write_buffers.get_mut(thread) {
+                    buffer.insert(x.to_string(), value);
+                } else {
+                    self.memory.insert(x.to_string(), value);
+                }
             }
         }
     }
 
     pub fn write_local(&mut self, thread: &str, x: &str, value: u32) {
-        self.memory.insert(format!("{thread}.{x}"), value);
+        match self.memory_model {
+            MemoryModel::Sc => {
+                self.memory.insert(format!("{thread}.{x}"), value);
+            }
+            MemoryModel::Tso => {
+                if let Some(buffer) = self.write_buffers.get_mut(thread) {
+                    buffer.insert(format!("{thread}.{x}"), value);
+                } else {
+                    //
+                }
+            }
+        }
+    }
+
+    pub fn read(&mut self, x: &str, thread: &str) -> u32 {
+        match self.memory_model {
+            MemoryModel::Sc => {
+                self.memory.get(x).copied().unwrap_or(0)
+            }
+            MemoryModel::Tso => {
+                if let Some(buffer) = self.write_buffers.get_mut(thread) {
+                    if buffer.contains_key(x) {
+                        buffer.get(x).copied().unwrap_or(0)
+                    } else {
+                        self.memory.get(x).copied().unwrap_or(0)
+                    }
+                    
+                } else {
+                    // this happens when main thread executes
+                    self.memory.get(x).copied().unwrap_or(0)
+                }
+            }
+        }
+       
     }
 
-    pub fn read(&self, x: &str) -> u32 {
-        self.memory.get(x).copied().unwrap_or(0)
+    pub fn read_local(&mut self, thread: &str, x: &str) -> u32 {
+        match self.memory_model {
+            MemoryModel::Sc => {
+                self.read(format!("{thread}.{x}").as_str(), thread)
+            }
+            MemoryModel::Tso => {
+                if let Some(buffer) = self.write_buffers.get_mut(thread) {
+                    if buffer.contains_key(format!("{thread}.{x}").as_str()) {
+                        buffer.get(format!("{thread}.{x}").as_str()).copied().unwrap_or(0)
+                    } else {
+                        self.read(format!("{thread}.{x}").as_str(), thread)
+                    }
+                } else {
+                    self.read(format!("{thread}.{x}").as_str(), thread)
+                }
+            }
+        }
```

Comment:
No special logic should be needed here since reads should just read from the current memory and especially not from the write buffers.